agreement/pseudonode.go

```diff
@@ -193,33 +195,57 @@ func (n asyncPseudonode) MakeVotes(ctx context.Context, r round, p period, s ste
 	}
 }
 
+// load the participation keys from the account manager ( as needed ) for the
+// current round.
+func (n *asyncPseudonode) loadRoundParticipationKeys(voteRound basics.Round) {
+	// if we've already loaded up the keys, then just skip loading them.
+	if n.participationKeysRound == voteRound {
+		return
+	}
+
+	cparams, err := n.ledger.ConsensusParams(ParamsRound(voteRound))
+	if err != nil {
+		// if we cannot figure out the balance round number, reset the parameters so that we won't be sending
+		// any vote.
+		n.log.Warnf("asyncPseudonode: unable to retrieve consensus parameters for voting round %d : %v", voteRound, err)
+		n.participationKeysRound = basics.Round(0)
+		n.participationKeys = nil
+		return
+	}
+	balanceRound := balanceRound(voteRound, cparams)
+
+	// otherwise, we want to load the participation keys.
+	n.participationKeys = n.keys.VotingKeys(voteRound, balanceRound)
+	n.participationKeysRound = voteRound
+}
+
 func (n asyncPseudonode) makeProposalsTask(ctx context.Context, r round, p period) pseudonodeProposalsTask {
-	participation := n.getParticipations("asyncPseudonode.makeProposalsTask", r)
+	n.loadRoundParticipationKeys(r)
 
 	pt := pseudonodeProposalsTask{
 		pseudonodeBaseTask: pseudonodeBaseTask{
 			node:          &n,
 			context:       ctx,
-			participation: participation,
+			participation: n.participationKeys,
```

Comment:
It's sections like this that make me a little uneasy. `n.participationKeys` is always available, so someone unfamiliar with the code might mistakenly use stale data.

Previously you were required to call the function, now the developer must remember to update the cache.

Reply:
So, the only way to work around that (nicely) would be to create a caching object that would cache the data, and return a slice. While this is clearly doable, I think that for this use case, it's definitely crossing the line to the over engineering realm. ( i.e. it's an internal functionality within an object; the object is fairy simple; the usage is very localized and have exactly two instance )

Reply:
Maybe I'm not explaining myself properly, I was thinking it would simply be a helper function and allow the existing code to work with the improved algorithm. Effectively this:

```
func (n asyncPseudonode) getParticipations(procName string, round basics.Round) []account.Participation {
    n.loadRoundParticipationKeys(r)
    return n.participationKeys
}
```

Reply:
I honestly don't see how adding this function would help - If I were to read this, I would probably think of that as a `spaghetti code`. But Iets see the other reviewer's for their view on that ? @brianolson @nicholasguoalgorand

Reply:
I kinda like the accessor that ensures cache state. If implemented, add a comment to the variable that it should be read from the accessor.

Reply:
Why are we using the logic of setting `n.participationKeys` in `loadRoundParticipationKeys` when we call `loadRoundParticipationKeys` every time before accessing `n.participationKeys` anyway? Could we have `loadRoundParticipationKeys` just return the list of keys?